AppTaskQueue/appscale/taskqueue/queue.py

```diff
@@ -235,7 +235,538 @@ def __repr__(self):
     return '<PushQueue {}: {}>'.format(self.name, attr_str)
 
 
+class PostgresPullQueue(Queue):
+
+  """
+  Before using Postgres implementation, make sure that
+  connection using appscale user can be created:
+  /etc/postgresql/9.5/main/pg_hba.conf
+  """
+
+  def __init__(self, queue_info, app, pg_connection=None):
+    """ Create a PostgresPullQueue object.
+
+    Args:
+      queue_info: A dictionary containing queue info.
+      app: A string containing the application ID.
+      pg_connection: A psycopg2 connection to PostgreSQL.
+    """
+    super(PostgresPullQueue, self).__init__(queue_info, app)
+    self.pg_connection = pg_connection
+    self.pg_cursor = pg_connection.cursor()
+    self.ensure_tables_created()
+
+  @property
+  def tasks_table_name(self):
+    return 'pullqueue-{}'.format(self.name)
+
+  def ensure_tables_created(self):
+    try:
+      self.pg_cursor.execute(
+        'CREATE TABLE IF NOT EXISTS "{table_name}" ('
+        '  task_name varchar(500) NOT NULL,'
+        '  time_enqueued timestamp NOT NULL,'
+        '  retry_count integer NOT NULL,'
+        '  lease_expires timestamp NOT NULL,'
+        '  payload text,'
+        '  tag varchar(500),'
+        '  PRIMARY KEY (task_name)'
+        ');'
+        'CREATE INDEX "{table_name}-eta-retry-tag-index" '
+        '  ON "{table_name}" USING BTREE (lease_expires, retry_count, tag);'
+        'CREATE INDEX "{table_name}-retry-eta-tag-index" '
+        '  ON "{table_name}" (retry_count, lease_expires, tag);'
+        .format(table_name=self.tasks_table_name)
+      )
+      self.pg_connection.commit()
+    except Exception as err:
+      logger.error('Rolling back transaction ({err})'.format(err=err))
+      self.pg_connection.rollback()
+      raise
+
+  def add_task(self, task):
+    """ Adds a task to the queue.
+
+    Args:
+      task: A Task object.
+    Raises:
+      InvalidTaskInfo if the task ID already exists in the queue
+        or it doesn't have payloadBase64 attribute.
+    """
+    import psycopg2
+    if not hasattr(task, 'payloadBase64'):
+      raise InvalidTaskInfo('{} is missing a payload.'.format(task))
+
+    enqueue_time = datetime.datetime.utcnow()
+    try:
+      lease_expires = task.leaseTimestamp
+    except AttributeError:
+      lease_expires = datetime.datetime.utcfromtimestamp(0)
+
+    try:
+      task.queueName = self.name
+      task.enqueueTimestamp = enqueue_time
+      task.leaseTimestamp = lease_expires
+
+      self.pg_cursor.execute(
+        'INSERT INTO "{table}" ( '
+        '  task_name, payload, time_enqueued, '
+        '  lease_expires, retry_count, tag '
+        ')'
+        'VALUES ( '
+        '  %(task_name)s, %(payload)s, %(time_enqueued)s, '
+        '  %(lease_expires)s, %(retry_count)s, %(tag)s '
+        ')'.format(table=self.tasks_table_name),
+        vars={
+          'task_name': task.id,
+          'payload': task.payloadBase64,
```

Comment:
I believe I made a mistake by storing the data as base64-encoded strings in the Cassandra implementation. The base64-encoding should be left as a serialization detail for the REST API layer only.

Reply:
I'd like to keep this PR focused on Postgres implementation of Pull queues and would avoid changes related to all regular Push and Pull queues.

Reply:
Sorry, I was not suggesting that you change existing implementations. It would just be a detail about how Postgres is storing the payload. You could just decode the base64 string right before storing it. In practice, it would leave behind an unnecessary encode step, but that would be easier to fix later than having to worry about converting existing data.